### PR TITLE
 report a error when the dest to redirection url does not match any files's publish URL, but the redirect_with_id flag has been set as true

### PR DESCRIPTION
**Pull request by 928PJY**
https://dev.azure.com/ceapex/Engineering/_workitems/edit/112321

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5005)

### CODE REVIEW

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.Graph;
```
Comment (live1206) [inline src/docfx/build/redirection/RedirectionMap.cs]: remove

File: src/docfx/Errors.cs [hunk 2]
```diff
@@ -43,6 +43,12 @@ public static Error RedirectionUrlInvalid(SourceInfo<string> source)
         public static Error RedirectionUrlConflict(SourceInfo<string> source)
             => new Error(ErrorLevel.Warning, "redirection-url-conflict", $"The '{source}' appears twice or more in the redirection mappings", source);
 
+        /// <summary>
+        /// The dest to redirection url does not match any files's publish URL, but the redirect_with_id flag has been set as true
+        /// </summary>
+        public static Error RedirectionUrlNotExisted(SourceInfo<string> source)
+            => new Error(ErrorLevel.Error, "redirection-url-not-existed", $"The redirect url '{source}' does not match any file's publish URL, but the redirect_with_id flag has been set as true", source);
```
Comment (live1206) [inline src/docfx/Errors.cs]: it says report `warning` in https://dev.azure.com/ceapex/Engineering/_workitems/edit/112321, upgraded to `error`?

File: src/docfx/build/redirection/RedirectionMap.cs (hunk 1 above)
Comment (yufeih) [inline src/docfx/build/redirection/RedirectionMap.cs]: ~~using Microsoft.Graph;~~

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 3]
```diff
@@ -54,12 +56,11 @@ public static RedirectionMap Create(ErrorLog errorLog, Docset docset, Func<strin
 
             var redirectionsBySourcePath = redirections.ToDictionary(file => file.FilePath.Path, PathUtility.PathComparer);
 
+            CheckInvalidRedrectUrl();
             return new RedirectionMap(redirectionsBySourcePath, redirectionsByRedirectionUrl);
 
             void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool redirectDocumentId = false)
```
Comment (yufeih) [inline src/docfx/build/redirection/RedirectionMap.cs]: nit: consider turn this into a member function, this method is getting big enough

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 4]
```diff
@@ -111,6 +113,24 @@ void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool redirect
                     }
                 }
             }
+
+            void CheckInvalidRedrectUrl()
+            {
+                var redirectUrls = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+                var publishUrls = buildFiles.Select(file => file.SiteUrl).Union(redirections.Select(redirection => redirection.SiteUrl)).ToHashSet();
```
Comment (yufeih) [inline src/docfx/build/redirection/RedirectionMap.cs]: Union creates a hashset internally, there is already a ToHashSet at the end, so Concat is faster in this case
```suggestion
                var publishUrls = buildFiles.Select(file => file.SiteUrl).Concat(redirections.Select(redirection => redirection.SiteUrl)).ToHashSet();
```

File: src/docfx/Errors.cs (hunk 2 above)
Comment (928PJY) [reply src/docfx/Errors.cs]: Yes, since it should break the build of this file, so upgraded to `error`

**Pull request merged**